### ISSUE TITLE
Как убить кэш у bem-tools?

### ISSUE THREAD

**Issue by pavelpower** [bem/bem-forum-content-ru#167]
Смотря на то как написаны COA нет ясности, а из API то как мне использовать команды?

Вот мне надо собирать bem-tools-ом так, чтобы он не использовал cache, 

И вроде нечто такое есть `forceCache: ture`, но как мне 

```javascript
require('bem').build({ forceCahse: true, //.....  }) 
````

выполнить то? Чтобы не использовал он Cache.... А то подает с ним постоянно.

**Comment by zxqfox** [comment]
https://github.com/express-bem/tools-make/blob/master/lib/index.js#L55
там два параметра. один — опции, второй — аргументы

**Comment by tadatuta** [comment]
@pavelpower скорее всего тебе нужно что-то типа:

```js
require('bem').api.make({ verbosity : 'error' }, { targets : 'path/to/target' })
    .then(function() {
        require('bem/lib/level').resetLevelsCache();
    });
```

**Comment by zxqfox** [comment]
@tadatuta т.е. https://github.com/express-bem/tools-make/blob/master/lib/index.js#L54 не в том месте стоит? надо после then?

**Comment by pavelpower** [comment]
```javascript
...
"devDependencies": {
    "bem": "~0.9.0",
...
```

```javascript

require('bem').api.make({ verbosity : 'error' }, { targets : 'path/to/target' })
    .then(function() {
        require('bem/lib/level').resetLevelsCache();
    });
```

**Comment by pavelpower** [comment]
версию менять сейчас не комильфо

```javacript
bem.build({
//...
})
```

**Comment by tadatuta** [comment]
@pavelpower либо я не понял, что ты хочешь сказать, либо версия у тебя актуальная.

@zxqfox Кроме того, что дергать нужно в `then()`, есть еще история про параллельные запросы (можно зацепить кэш незавершенной сборки и все взорвется). В качестве референса см. код `bem server`: https://github.com/bem/bem-tools/blob/support/0.9.x/lib/base-server.js#L188-L190 либо пример из дикой природы:
```js
var rebuildsInProgress = 0;

app
    .disable('x-powered-by')
    .enable('trust proxy')
    .use(express.logger(isProduction ? '' : 'dev'))
    .use(express.compress())
    .use(function(req, res, next) {
        if(isProduction || !/^\/_(.*)\.(css|js)$/.test(req.url)) return next();

        rebuildsInProgress++;

        BEM.api.make({ verbosity : 'error' }, { targets : staticDir + req.url })
            .then(function() {
                rebuildsInProgress--;
                rebuildsInProgress || require('bem/lib/level').resetLevelsCache();

                next();
         });
    })
    .use(express.static(pathToStatic, { maxAge : constants.cacheTime }))
    .use(express.bodyParser())
    .use(express.methodOverride())
    .use(csrf())
    .use(slashes());
```

**Comment by zxqfox** [comment]
@tadatuta А почему именно в then? Типа, поел — убери за собой?

Кстати, этот код будет падать в кластере, да?

**Comment by tadatuta** [comment]
призову @scf2k, возможно он вспомнит подробности

**Comment by scf2k** [comment]
Про какой кеш речь?

**Comment by pavelpower** [comment]
замутили мы параллельно запускать несколько сборок под gulp с помощью bem-tools
и все бы хорошо, но в корневой папке `.bem` постоянно появляется `.bem/cache` и он всю прекрасную схему ломает.
а вот если удалять постоянно это `cache` то все как бы работает нормально. Да, и при изменениях файлов в сборку попадают изменения.
Так вот, как сделать так чтобы не было сего cacha сразу?

**Comment by zxqfox** [comment]
`ln -s /dev/null .bem/cache`? :-D

**Comment by pavelpower** [comment]
ееееее, вот что значит мыслить ширшее  :)

**Comment by apsavin** [comment]
А как вы из-под галпа bem-tools  запускаете? Типа `require('bem').api.make`?
Если вы запускаете для нескольких разных проектов, то почему папка `.bem/cache` одна?

**Comment by apsavin** [comment]
Так-то от кэша отказываться не оч круто, это замедляет сборку.

**Comment by pavelpower** [comment]
Если по чеснаку, то что с кешем что без него все едино.

А собираем сразу под несколько бандлов одновременно, с разными настройками по техналогиям и прочей ересью.

Ну вот мне допустим надо сделать сборку сразу под 100500 устройств, да при этом, у samsung в сборку надо подтягивать спец файлы которые не укладываются в БЭМ архитектуру. По сему, такие вещи проще делать на синергии gulp и БЭМ.

**Comment by apsavin** [comment]
Ну, если в настройках указать, что можно кешировать bem-core и bem-components, то кеш дает прирост скорости.

То есть проект один и вы сборку запускаете несколько раз, чтобы параллельно собирались бандлы?

**Comment by zxqfox** [comment]
У меня принципиально новая идея: а почему не хранить кеш в редисе?

**Comment by tadatuta** [comment]
@pavelpower кто понял жизнь, тот не спешит? ;)

**Comment by pavelpower** [comment]
@tadatuta стал философом - достиг просветления!

@zxqfox - и уж точно надо опционально. А еще хорошо бы указывать куда ложить кэш, причем для каждого бандла свой кэш - это то, что мне сейчас нужно.

**Comment by pavelpower** [comment]
_небольшая ремарка, у меня несколько слоев переопределения у которых есть одинаковые по названию бандлы, и при сборке сразу всех кэш жутко конфликтует_

**Comment by tadatuta** [comment]
@pavelpower меня не покидает ощущение, что вы сами себе создаете проблемы.
можешь подробно рассказать про верхнеуровневую задачу, которую вы хотите решить? возможно, я смогу посоветовать какой-то более простой путь.

**Comment by pavelpower** [comment]
@tadatuta ` require('bem/lib/level').resetLevelsCache();` заработал только после модификации

```javascript
exports.resetLevelsCache = function(all) {
    for(var l in levelCache) {
        var level = levelCache[l];

        //if (!level.cache || all) level.cache = null;
        level.cache = false;
        level.file = null 
    }
};
```

**Comment by zxqfox** [comment]
@tadatuta В bem-server залипает кеш на фейлах, похоже.

https://github.com/bem/bem-tools/blob/support/0.9.x/lib/base-server.js#L201

`_this._targetsInProcess` растет всегда, а снижается только при успешных рендрах

**Comment by tadatuta** [comment]
@zxqfox не возьмусь утверждать, но вроде как декремент вызывается на `.fin()` — это метод `q`, который дергается в любом случае, даже если промис зареджектился или произошло исключение

**Comment by zxqfox** [comment]
@tadatuta Да, ступил. Если `.fin` вызывается всегда гарантированно — то проблем с этим нет. Правда, не факт, что это работает ок под виндой, т.к. там были с этим проблемки.
